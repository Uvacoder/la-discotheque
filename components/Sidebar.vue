<template>
  <div class="Sidebar">
    <Logo />
    <ul class="Sidebar__Container">
      <li class="Sidebar__Group">
        <nuxt-link
          class="Sidebar__Link"
          tag="div"
          to="/collection" >
          <a>Collection</a>
        </nuxt-link>
        <nuxt-link
          class="Sidebar__Link"
          tag="div"
          to="/wishlist" >
          <a>Wishlist</a>
        </nuxt-link>
      </li>
      <nuxt-link
        class="Sidebar__Link"
        tag="li"
        to="/about" >
        <a>About</a>
      </nuxt-link>
    </ul>
  </div>
</template>

<script>
import Logo from '~/components/Logo.vue'

export default {
  components: {
    Logo,
  }
}
</script>

<style>
.Sidebar {
  position: fixed;
  width: 96px;
  height: 100%;
  left: 0;
  top: 0;
  bottom: 0;
  padding-top: 40px;
}

.Sidebar__Container {
  display: flex;
  height: calc(100% - 140px);
  flex-flow: row wrap;
  align-items: center;
}

.Sidebar__Link {
  font-family: "IBMPlexMono-Regular";
  width: 100%;
  text-align: center;
  line-height: 96px;
  padding: 1em 0;
  position: relative;
}

.Sidebar__Link::after {
    content: '';
    display: block;
    width: 8px;
    height: 8px;
    background: #0053FF;
    border-radius: 4px;
    position: absolute;
    right: -4px;
    top: 0;
    bottom: 0;
    margin: auto;
    visibility: hidden;
}

.Sidebar__Link.nuxt-link-active {
  font-family: "IBMPlexMono-SemiBold";
  color: #0053FF;
  width: 100%;
  text-align: center;
  line-height: 96px;
}
.Sidebar__Link.nuxt-link-active:after {
    visibility: visible;
}

.Sidebar__Link:last-of-type {
  font-family: "IBMPlexMono-SemiBold";
  align-self: flex-end;
}

.Sidebar__Link a {
  display: block;
  color: inherit;
  text-decoration: none;
  transform: rotate(-90deg);
  transform-origin: center;
}

.Sidebar ul {
  list-style-type: none;
  padding: 0;
}

.Sidebar .Logo {
  margin-left: 40px;
}
</style>
