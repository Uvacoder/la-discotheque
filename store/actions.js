require('isomorphic-fetch');
// We declare it here because otherwise it will be overriden by the build process.
const BASE_URL = process.env.BASE_URL || 'http://localhost:3000';

const actions = {
  'GET_COLLECTION': async ({ commit, state }) => {
    const query = await fetch(`${state.BASE_URL}/api/collection`);
    const payload = await query.json();

    if (payload.ok) {
      return commit('SET_COLLECTION', payload)
    }
    // TOOD handle error
  },
  'GET_WANTLIST': async ({ commit, state }) => {
    const query = await fetch(`${state.BASE_URL}/api/wantlist`);
    const payload = await query.json();

    if (payload.ok) {
      return commit('SET_WANTLIST', payload)
    }
    // TOOD handle error
  },
  'GET_RELEASE': async ({ getters, state, commit }, id) => {
    const query = await fetch(`${state.BASE_URL}/api/releases/${id}`);
    const payload = await query.json();

    if (payload.ok) {
      return commit('SET_RELEASE', payload)
    }
    // TOOD handle error
  },

  async nuxtServerInit({ commit, dispatch }, { route }) {
    // TODO improve checks
    commit('SET_BASE_URL', BASE_URL)
    if (route.name == "collection-slug") {
      dispatch('GET_COLLECTION');
    }
    else if (route.name == "wishlist-slug") {
      dispatch('GET_WANTLIST');
    }
  }
}

export default actions
