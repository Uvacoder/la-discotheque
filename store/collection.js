import fetch from 'isomorphic-fetch'

export const state = () => ({
  items: [],
  isLoaded: false
})

export const mutations = {
  SET_COLLECTION(state, payload) {
    state.items = [...payload.data.releases]
    state.isLoaded = true
  }
}

export const getters = {
  count(state) {
    return state.items.length
  }
}

export const actions = {
  GET_COLLECTION: async ({ commit, rootState }) => {
    const query = await fetch(`${rootState.BASE_URL}/api/collection`)
    const payload = await query.json()

    if (payload.ok) {
      return commit('SET_COLLECTION', payload)
    }
    // TODO handle error
  }
}
