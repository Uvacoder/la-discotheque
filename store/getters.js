const getters = {
  getItemById: ( state ) => (folderName, id) => {
    if (!folderName) {
      console.error('[store][getters] getItemById - missing arg folderName');
      return;
    }

    folderName = folderName === "wishlist" ? "wantlist" : folderName;

    const folder = state[folderName];
    if (folder.items && folder.items.length > 0) {
      return folder.items.filter( item => item.id == id)[0]
    }
    return false;
  },

  getItemByIndex: ( state ) => (folderName, index) => {
    if (!folderName) {
      console.error('[store][getters] getItemByIndex - missing arg folderName');
      return;
    }

    folderName = folderName === "wishlist" ? "wantlist" : folderName;

    const folder = state[folderName];
    if (folder.items && folder.items.length > 0) {
      return folder.items[index]
    }
    return false;
  },

  getFolderIndexById: ( state ) => (folderName, id) => {
    if (!folderName) {
      console.error('[store][getters] getFolderIndexById - missing arg folderName');
      return;
    }

    folderName = folderName === "wishlist" ? "wantlist" : folderName;

    const folder = state[folderName];
    if (folder.items && folder.items.length > 0) {
      return folder.items.findIndex( item => item.id == id)
    }
    return false;
  },

  collectionItemsCount: (state) => {
    return state.collection.items.length
  },

  wantlistItemsCount: (state) => {
    return state.wantlist.items.length
  }
}

export default getters
