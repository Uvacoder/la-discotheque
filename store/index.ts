import { ActionContext, Commit, Dispatch } from 'vuex'
import { getAccessorType } from 'typed-vuex'
import { Context } from '@nuxt/types'

// import Vuex from 'vuex'
// import fetch from 'isomorphic-fetch'

const BASE_URL = process.env.BASE_URL || 'http://localhost:3000'

export const state = () => ({
  BASE_URL: '',
  currentRelease: {},
  sidebarOpened: false,
  isMobile: false
})

export type RootState = ReturnType<typeof state>

export const mutations = {
  SET_BASE_URL(state: RootState, url: string) {
    state.BASE_URL = url
  },
  SET_RELEASE(state: RootState, release: Object) {
    state.currentRelease = Object.assign({}, release.data)
  },
  SET_SIDEBAR_VISIBLE(state: RootState, visible: boolean) {
    state.sidebarOpened = visible
  },
  SET_LAYOUT_MOBILE(state: RootState, isMobile: boolean) {
    state.isMobile = isMobile
  }
}

export const getters = {
  getItemById: (state: RootState) => (folderName: String, id: number) => {
    // if (!folderName) {
    //   console.error('[store][getters] getItemById - missing arg folderName')
    //   return
    // }

    folderName = folderName === 'wishlist' ? 'wantlist' : folderName

    const folder = state[folderName]
    if (folder.items && folder.items.length > 0) {
      return folder.items.filter((item) => item.id === id)[0]
    }
    return false
  },

  getItemByIndex: (state: RootState) => (folderName: String, index: number) => {
    folderName = folderName === 'wishlist' ? 'wantlist' : folderName

    const folder = state[folderName]
    if (folder.items && folder.items.length > 0) {
      return folder.items[index]
    }
    return false
  },

  getFolderIndexById: (state: RootState) => (
    folderName: String,
    id: number
  ) => {
    folderName = folderName === 'wishlist' ? 'wantlist' : folderName

    const folder = state[folderName]
    if (folder.items && folder.items.length > 0) {
      return folder.items.findIndex((item) => item.id === id)
    }
    return false
  }
}

export const actions = {
  GET_RELEASE: async (
    { state, commit }: ActionContext<RootState, RootState>,
    id: number
  ) => {
    const query = await fetch(`${state.BASE_URL}/api/releases/${id}`)
    const payload = await query.json()

    if (payload.ok) {
      return commit('SET_RELEASE', payload)
    }
    // TODO handle error
  },
  // async nuxtServerInit(
  //   { commit, dispatch }: ActionContext<Commit, Dispatch>,
  //   { route }: Context
  // ) {
  //   // TODO improve checks
  //   await commit('SET_BASE_URL', BASE_URL)
  //   if (route.name === 'collection-slug') {
  //     await dispatch('collection/GET_COLLECTION')
  //   } else if (route.name === 'wishlist-slug') {
  //     await dispatch('wantlist/GET_WANTLIST')
  //   }
  // }
}

export const accessorType = getAccessorType({
  state,
  getters,
  mutations,
  actions
})
