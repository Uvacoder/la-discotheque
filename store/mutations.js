const mutations = {
  'SET_BASE_URL': (state, payload) => {
    state.BASE_URL = payload;
  },
  'SET_COLLECTION': (state, payload) => {
    state.collection.items = [...payload.data.releases]
    state.collection.isLoaded = true;
  },
  'SET_WANTLIST': (state, payload) => {
    state.wantlist.items = [...payload.data.wants]
    state.wantlist.isLoaded = true;
  },
  'SET_RELEASE': (state, payload) => {
    state.currentRelease = Object.assign({}, payload.data)
  }
}

export default mutations
