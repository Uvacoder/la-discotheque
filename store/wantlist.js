import fetch from 'isomorphic-fetch'

export const state = () => ({
  items: [],
  isLoaded: false
})

export const mutations = {
  SET_WANTLIST(state, payload) {
    state.items = [...payload.data.wants]
    state.isLoaded = true
  }
}

export const getters = {
  count(state) {
    return state.items.length
  }
}

export const actions = {
  GET_WANTLIST: async ({ commit, rootState }) => {
    const query = await fetch(`${rootState.BASE_URL}/api/wantlist`)
    const payload = await query.json()

    if (payload.ok) {
      return commit('SET_WANTLIST', payload)
    }
    // TODO handle error
  }
}
