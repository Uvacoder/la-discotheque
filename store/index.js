import Vuex from 'vuex'

import getters from './getters'
import mutations from './mutations'
import actions from './actions'

const createStore = () => {
  return new Vuex.Store({
    state: {
      BASE_URL: '',
      currentRelease: {},
      collection: {
        items: [],
        isLoaded: false
      },
      wantlist: {
        items: [],
        isLoaded: false
      },
      sidebarOpened: false,
      isMobile: false
    },
    getters,
    mutations,
    actions
  })
}

export default createStore
