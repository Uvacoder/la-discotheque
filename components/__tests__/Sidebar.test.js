import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'

import { state, mutations, getters, actions } from '~/store'
import Sidebar from '~/components/Sidebar.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

function createStore() {
  return new Vuex.Store({
    state,
    mutations,
    getters,
    actions
  })
}

describe('Sidebar.vue', () => {
  let store
  beforeEach(() => {
    store = createStore()
    Sidebar.components = Object.assign(Sidebar.components || {})
  })

  it('matches snapshot', () => {
    const wrapper = shallowMount(Sidebar, { store, localVue })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
