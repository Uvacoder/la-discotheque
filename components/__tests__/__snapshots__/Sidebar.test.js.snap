// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar.vue matches snapshot 1`] = `
<div role="navigation" class="Sidebar">
  <logo-stub></logo-stub>
  <ul class="Sidebar__Container">
    <li class="Sidebar__Group"><a tag="div" to="/collection" class="Sidebar__Link"><a>Collection</a></a> <a tag="div" to="/wishlist" class="Sidebar__Link"><a>Wishlist</a></a></li> <a tag="li" to="/about" class="Sidebar__Link"><a>About</a></a>
  </ul>
</div>
`;
